File: lua/glacier/widget/systray.lua
local Log = require("pinnacle.log") ---@diagnostic disable-line:unused-local
local signal = require("glacier.widget.signal")

local Widget = require("snowcap.widget")

local Base = require("glacier.widget.base")
local widget_signal = require("glacier.widget.signal")
local dbus = require("glacier.misc.dbus")

local SNIHost = require("glacier.protocols.status_notifier.host")

local Popup = require("snowcap.popup")
local Menu = require("glacier.menu")

---@class glacier.widget.systray
---@field mt metatable This module metatable
local systray = { mt = {} }

---@lcat nodoc
---@package
---
---Private segment of the systray module
---@class glacier.widget._systray
local _systray = {}

---@class glacier.widget.systray.MenuEntry: glacier.menu.entry.EntryBase
---@field private _item glacier.status_notifier.host.Item
---@field private _node glacier.status_notifier.LayoutNode
---@field private _menu_config glacier.menu.Config
local MenuEntry = Menu.entry.EntryBase:new_type()
MenuEntry.__index = MenuEntry
MenuEntry.__name = "glacier.widget.systray.MenuEntry"

function MenuEntry:label()
    return self._node:label()
end

function MenuEntry:disabled()
    return not self._node:is_enabled()
end

function MenuEntry:activate(hover)
    self._item:hover(self._node:id())

    if hover and self._node:is_submenu() then
        return Menu.action.entry.OpenMenu()
    end
end

function MenuEntry:submit()
    if self._node:is_submenu() then
        return Menu.action.entry.OpenMenu()
    else
        self._item:click(self._node:id())
        return Menu.action.menu.Close()
    end
end

function MenuEntry:open_menu()
    if not self._node:is_submenu() then
        return nil
    end

    local entries = {}
    local prev_is_sep = true
    local children = self._node:children()

    for _, node in ipairs(children) do
        if not node:is_visible() then
            goto continue
        end

        if node:is_separator() then
            if prev_is_sep then
                goto continue
            end

            table.insert(entries, Menu.entry.separator())
            prev_is_sep = true
        elseif node:is_standard() then
            prev_is_sep = false

            local menu_config = require("snowcap.util").deep_copy(self._menu_config)
            table.insert(entries, MenuEntry.new(self._item, node, menu_config))
        end

        ::continue::
    end

    -- We don't want to have the entries in the main config here
    local menu_config = require("snowcap.util").deep_copy(self._menu_config)
    menu_config.entries = entries
    return Menu.Menu:new(menu_config)
end

---@param active boolean
---@param style glacier.menu.entry.Style
---@param icon_mask glacier.image.AlphaMask
---@return snowcap.widget.WidgetDef
function MenuEntry:view_toggle(active, style, icon_mask)
    local state = Menu.entry.get_state(self, active)
    local entry_style = Menu.entry.entry_style_for_state(style, state)

    local label_widget = Widget.text({
        text = self:label(),
        width = Widget.length.Fill,
        style = {
            color = entry_style.fg_color,
            font = style.font,
            pixels = style.font_size,
        },
    })

    local icon_handle =
        icon_mask:to_image_handle(style.menu_indicator.color or entry_style.fg_color)
    local menu_icon = Widget.Image({
        handle = icon_handle,
        content_fit = Widget.image.content_fit.SCALE_DOWN,
        height = Widget.length.Fixed(16),
        width = Widget.length.Fixed(16),
    })

    return Widget.container({
        child = Widget.row({
            children = {
                label_widget,
                menu_icon,
            },
            item_alignment = Widget.alignment.CENTER,
        }),
        clip = true,
        padding = entry_style.padding,
        height = entry_style.height,
        width = Widget.length.Fill,
        valign = Widget.alignment.CENTER,
        style = {
            background_color = entry_style.bg_color,
            border = entry_style.border,
        },
    })
end

function MenuEntry:view(active, style)
    local _icons = require("glacier.misc.icons")

    if self._node:is_submenu() then
        return Menu.entry.default_menu_view(self, active, style)
    elseif self._node:is_checkbox() then
        return self:view_toggle(active, style, _icons.checkbox.select(self._node:is_toggled()))
    elseif self._node:is_radio() then
        return self:view_toggle(active, style, _icons.radio.select(self._node:is_toggled()))
    else
        return Menu.entry.default_entry_view(self, active, style)
    end
end

function MenuEntry.new(item, node, menu_config)
    return MenuEntry:super({
        _key = nil,
        _item = item,
        _node = node,
        _menu_config = menu_config,
    })
end

---Action to execute upon update.
---
---@class glacier.widget.systray.MessageAction
---@field action glacier.widget.systray.Action
---@field item string

---@enum glacier.widget.systray.Action
_systray.Action = {
    ACTIVATE = "systray::activate_item",
    DEACTIVATE = "systray::deactivate_item",
    ENTER = "systray::enter_item",
    EXIT = "systray::exit_item",
    ---@param item string
    ---@return glacier.widget.systray.MessageAction
    Activate = function(item)
        return { action = _systray.Action.ACTIVATE, item = item }
    end,
    ---@param item string
    ---@return glacier.widget.systray.MessageAction
    Deactivate = function(item)
        return { action = _systray.Action.DEACTIVATE, item = item }
    end,

    ---@param item string
    ---@return glacier.widget.systray.MessageAction
    Enter = function(item)
        return { action = _systray.Action.ENTER, item = item }
    end,

    ---@param item string
    ---@return glacier.widget.systray.MessageAction
    Exit = function(item)
        return { action = _systray.Action.EXIT, item = item }
    end,
}

---@class glacier.widget.systray.IconStyle
---@field border? snowcap.widget.Border
---@field padding? snowcap.widget.Padding
---@field bg_color? snowcap.widget.Color

local _icon_style_keys = {
    border = 1,
    padding = 1,
    bg_color = 1,
}

---@class glacier.widget.systray.Style
---@field bg_color? snowcap.widget.Color Background color for the whole systray.
---@field border? snowcap.widget.Border Border around the whole systray.
---@field spacing? number Spacing between icons.
---@field padding? snowcap.widget.Padding Padding for the whole row of icons.
---@field active? glacier.widget.systray.IconStyle Style
---@field hovered? glacier.widget.systray.IconStyle
---@field active_hovered? glacier.widget.systray.IconStyle Style
---@field default? glacier.widget.systray.IconStyle

---@return glacier.widget.systray.Style
function systray.default_style()
    ---@type glacier.widget.systray.Style
    return {
        bg_color = nil,
        border = nil,
        spacing = 1,
        padding = nil,
        active = nil,
        hovered = nil,
        default = {
            padding = {
                top = 2,
                left = 2,
                right = 2,
                bottom = 2,
            },
        },
    }
end

---@package
---@param style glacier.widget.systray.Style
---@return glacier.widget.systray.IconStyle
function _systray._get_active_style(style)
    if not style.active then
        return style.default
    end

    return {
        border = style.active.border or style.default.border,
        padding = style.active.padding or style.default.padding,
        bg_color = style.active.bg_color or style.default.bg_color,
    }
end

---@package
---@param style glacier.widget.systray.Style
---@return glacier.widget.systray.IconStyle
function _systray._get_hovered_style(style)
    if not style.hovered then
        return style.default
    end

    return {
        border = style.hovered.border or style.default.border,
        padding = style.hovered.padding or style.default.padding,
        bg_color = style.hovered.bg_color or style.default.bg_color,
    }
end

---@package
---@param style glacier.widget.systray.Style
---@return glacier.widget.systray.IconStyle
function _systray._get_active_hovered_style(style)
    local active = style.active or {}
    local hovered = style.hovered or {}
    local active_hovered = style.active_hovered or {}

    ---@type glacier.widget.systray.IconStyle
    local icon_style = {}

    for k, _ in pairs(_icon_style_keys) do
        icon_style[k] = active_hovered[k] or active[k] or hovered[k] or style.default[k]
    end

    return icon_style
end

---@package
---@param style glacier.widget.systray.Style
---@param active boolean
---@param hovered boolean
---@return glacier.widget.systray.IconStyle
function _systray.get_icon_style(style, active, hovered)
    if active and hovered then
        return _systray._get_active_hovered_style(style)
    elseif active then
        return _systray._get_active_style(style)
    elseif hovered then
        return _systray._get_hovered_style(style)
    else
        return style.default
    end
end

---@alias glacier.widget.systray.ViewFn fun(items: snowcap.widget.WidgetDef[], style:glacier.widget.systray.Style): snowcap.widget.WidgetDef

---@param children snowcap.widget.WidgetDef[]
---@param style glacier.widget.systray.Style
function systray.default_view(children, style)
    return Widget.container({
        child = Widget.row({
            height = Widget.length.Fill,
            item_alignment = Widget.alignment.CENTER,
            spacing = style.spacing,
            children = children,
        }),
        padding = style.padding,
        style = {
            border = style.border,
            background_color = style.bg_color,
        },
    })
end

---@alias glacier.widget.systray.IconViewFn fun(item: glacier.status_notifier.host.Item, style: glacier.widget.systray.IconStyle): snowcap.widget.WidgetDef?

---@param item glacier.status_notifier.host.Item
---@param style glacier.widget.systray.IconStyle
function systray.default_icon_view(item, style)
    local _icons = require("glacier.misc.icons")
    local _color = require("glacier.misc.color")
    local icon_name = item.icon_name

    local image_handle
    if icon_name and string.len(icon_name) > 0 then
        local path = item.icon_theme_path .. "/" .. icon_name .. ".png"
        image_handle = {
            path = path,
        }
    else
        local pixmap = item.icon_pixmap[1]

        if pixmap then
            image_handle = {
                rgba = { width = pixmap.x, height = pixmap.y, rgba = pixmap.data },
            }
        end
    end

    if not image_handle then
        local icon_mask = _icons.misc.broken_picture()
        image_handle = icon_mask:to_image_handle(_color.from_hex("#FFFFFF"))
    end

    return Widget.container({
        child = Widget.Image({
            handle = image_handle,
        }),
        padding = style.padding,
        style = {
            border = style.border,
            background_color = style.bg_color,
        },
    })
end

---@class glacier.widget.systray.SysTray: glacier.widget.Base
---@field private _host glacier.status_notifier.Host
---@field private _active string
---@field private _hovered string
---@field private _menu_config glacier.menu.Config
---@field private _menu? glacier.menu.Menu
---@field private _menu_signals? glacier.menu.MenuSignals
---@field private _view_fn? glacier.widget.systray.ViewFn
---@field private _icon_view_fn? glacier.widget.systray.IconViewFn
---@field private _style glacier.widget.systray.Style
local SysTray = Base:new_class({ type = "SysTray" })

function SysTray:_view_icon(key, item)
    local id = item.id

    local icon_style =
        _systray.get_icon_style(self._style, self._active == key, self._hovered == key)
    local icon_view
    if self._icon_view_fn then
        local ok
        ok, icon_view = pcall(self._icon_view_fn, item, icon_style)

        if not ok then
            Log.error(
                ("While calling view function for item %s: %s"):format(item.id, tostring(icon_view))
            )
        end
    else
        icon_view = systray.default_icon_view(item, icon_style)
    end

    if not icon_view then
        return nil
    end

    local marea = Widget.mouse_area({
        child = icon_view,
        on_release = {
            widget_id = self:id(),
            action = _systray.Action.Activate(key),
        },
        on_enter = {
            widget_id = self:id(),
            action = _systray.Action.Enter(key),
        },
        on_exit = {
            widget_id = self:id(),
            action = _systray.Action.Exit(key),
        },
    })

    return Widget.container({
        id = id,
        child = marea,
    })
end

function SysTray:view()
    local children = {}

    for k, item in pairs(self._host:items()) do
        local child = self:_view_icon(k, item)
        table.insert(children, child)
    end

    if self._view_fn then
        local ok, ret = pcall(self._view_fn, children, self._style)
        if not ok then
            Log.error(("While calling Systray's view function: %s"):format(tostring(ret)))
            return nil
        else
            return ret
        end
    end

    return systray.default_view(children, self._style)
end

---@param parent glacier.Surface
function SysTray:_activate_item(item_id, parent)
    local item = self._host:items()[item_id]

    item:about_to_show()

    if item.menu_tree then
        local root = MenuEntry.new(item, item.menu_tree, self._menu_config)
        local menu = root:open_menu()

        if menu then
            local parent_handle = parent:get_handle():as_parent()
            local ok, err = menu:show(parent_handle, nil, {
                position = Popup.position.AtWidget(item.id),
            })
            if not ok then
                Log.error(("While opening menu for %s: %s"):format(item.id, tostring(err)))
                return nil
            end

            self._menu = menu
            self._menu_signals = {
                closed = self._menu:connect(Menu.signal.CLOSED, function()
                    self:emit(signal.send_message, {
                        widget_id = self:id(),
                        action = _systray.Action.Deactivate(item_id),
                    })
                end),
            }

            self._active = item_id
        end
    end
end

---@param msg any
---@param parent glacier.Surface
function SysTray:update(msg, parent)
    if not msg then
        return
    end

    if msg.widget_id == self:id() then
        msg = msg.action --[[@as glacier.widget.systray.MessageAction]]

        if msg.action == _systray.Action.ACTIVATE then
            self:_activate_item(msg.item, parent)
        elseif msg.action == _systray.Action.DEACTIVATE then
            self._active = nil
            self._menu = nil
            self._menu_signals = nil
        elseif msg.action == _systray.Action.ENTER then
            self._hovered = msg.item
        elseif msg.action == _systray.Action.EXIT then
            if self._hovered == msg.item then
                self._hovered = nil
            end
        end
    end
end

function SysTray:refresh()
    self:emit(widget_signal.redraw_needed)
end

---@class glacier.widget.systray.Config
---@field menu_config? glacier.menu.Config
---@field style? glacier.widget.systray.Style
---@field view_fn? glacier.widget.systray.ViewFn
---@field icon_view_fn? glacier.widget.systray.IconViewFn

---@param config glacier.widget.systray.Config
---@return glacier.widget.systray.SysTray
function SysTray:new(config)
    config = config or {}

    ---@type glacier.widget.systray.Config
    local default_config = {
        menu_config = {
            direction = Menu.direction.DownLeft,
            popup_config = {
                anchor = Popup.anchor.BOTTOM_RIGHT,
                offset = { x = 0, y = 8 },
            },
            child_popup_config = {
                offset = { x = -2, y = 0 },
            },
        },
        style = systray.default_style(),
    }

    ---@type glacier.widget.systray.Config
    config = require("glacier.utils").merge_table(default_config, config)

    local ret = SysTray:super({
        _menu_config = config.menu_config,
        _style = config.style,
        _view_fn = config.view_fn,
        _icon_view_fn = config.icon_view_fn,
    })

    local id = ret:id()

    ret._host = SNIHost.new(dbus.session(), ("glacier-%d"):format(id), {
        on_register = function()
            ret:refresh()
        end,
        on_unregister = function()
            ret:refresh()
        end,
    })

    return ret
end

---@param ... glacier.widget.systray.Config
---@return glacier.widget.systray.SysTray
function systray.mt:__call(...)
    return SysTray:new(...)
end

return setmetatable(systray, systray.mt) --[[@as glacier.widget.systray]]

File: rust/glacier/src/widget/systray.rs
//! System tray widget.
//!
//! [`SysTray`] allows to display system trays icons, which display a menu when clicked. When the
//! view function is called, this widgets renders each icons, wrap their view in a [`MouseArea`]
//! which handles the click and hover. This first render pass can be overridden through the
//! [`icon_view_callback`] attribute, and default to calling [`SysTray::default_icon_view`].
//!
//! The resulting views are then passed to another function, which should render the list itself.
//! This second pass can be overridden by via the [`view_callback`] attribute, and default to
//! calling [`SysTray::default_view`].
//!
//! When an icon is clicked, a menu is opened relative to its containing widget. The specific
//! config used is held by the [`menu_config`] attribute, which default to
//! [`SysTray::default_menu_config`]. This default config expect the [`SysTray`] to be on the
//! right-hand side of a bar, opening the menu down and to the left.
//!
//! [`icon_view_callback`]: SysTray::icon_view_callback
//! [`view_callback`]: SysTray::view_callback
//! [`menu_config`]: SysTray::menu_config

use futures::prelude::*;
use std::marker::PhantomData;

use snowcap_api::{
    popup::{self, Position},
    widget::{
        self, Length,
        container::{self, Container},
        image::{self, Image},
        mouse_area::MouseArea,
        row::Row,
    },
};

use crate::{
    menu::{Entry, Menu, MenuConfig, PopupConfig, signal as menu_signal},
    protocols::status_notifier,
    signal::{self, WithEmitter},
    widget::{
        State, WeakState, Widget, WidgetMessage, WithState,
        base::WidgetBase,
        message::{self, MessageBuilder},
        signal as widget_signal,
        systray::style::default_style,
    },
};

pub mod menu;
pub mod style;
#[doc(inline)]
pub use style::{IconStyle, Style};

pub type Message = message::Message<Action>;
pub type ViewCallback<Msg> = Box<
    dyn Fn(Vec<widget::WidgetDef<Msg>>, &Style) -> Option<widget::WidgetDef<Msg>> + Send + Sync,
>;
pub type IconViewCallback<Msg> =
    Box<dyn Fn(&status_notifier::Item, IconStyle) -> Option<widget::WidgetDef<Msg>> + Send + Sync>;

/// [`SysTray`] inner state
pub struct Inner<Msg> {
    base: WidgetBase,
    host: status_notifier::Host,
    hovered: Option<String>,
    active: Option<String>,
    menu_config: MenuConfig,
    menu: Option<Menu<()>>,
    menu_signal: Option<signal::Handle<menu_signal::Closed>>,
    style: Style,
    view_callback: Option<ViewCallback<Msg>>,
    icon_view_callback: Option<IconViewCallback<Msg>>,
    message_builder: MessageBuilder<Action>,
    _msg: PhantomData<Msg>,
}

/// [`SysTray`] actions.
#[derive(Clone, Debug)]
pub enum Action {
    Activate(String),
    Deactivate(String),
    Enter(String),
    Exit(String),
}

/// Systray widget.
///
/// See module level [documentation] for more information.
///
/// [documentation]: self
#[derive(Clone)]
pub struct SysTray<Msg> {
    state: State<Inner<Msg>>,
}

/// Non-owning [`SysTray`].
#[derive(Clone)]
pub struct WeakSysTray<Msg>(WeakState<Inner<Msg>>);

impl<Msg> SysTray<Msg>
where
    Msg: Clone + Send + Sync + 'static,
{
    const WIDGET_TYPE: &'static str = "SysTray";

    /// Create a new [`SysTray`]
    pub async fn new(connection: &zbus::Connection) -> Self {
        let base = WidgetBase::new(Self::WIDGET_TYPE);
        let message_builder = MessageBuilder::new(base.id());
        let host = status_notifier::Host::new(connection, format!("glacier-{}", base.id())).await;

        let mut host_stream = host.item_stream().await;

        let state = State::new(Inner {
            base,
            host,
            hovered: None,
            active: None,
            menu_config: Self::default_menu_config(),
            menu: None,
            menu_signal: None,
            style: default_style(),
            view_callback: None,
            icon_view_callback: None,
            message_builder,
            _msg: PhantomData,
        });

        tokio::spawn({
            let weak = WeakSysTray(state.downgrade());
            async move {
                while host_stream.next().await.is_some() {
                    let Some(systray) = weak.upgrade() else {
                        break;
                    };

                    systray.emit(crate::widget::signal::RedrawNeeded);
                }
            }
        });

        Self { state }
    }

    /// Sets the configuration to use when spawning menus.
    pub fn menu_config(self, config: MenuConfig) -> Self {
        self.state.0.lock().unwrap().menu_config = config;

        self
    }

    /// Sets the [`SysTray`]'s style.
    pub fn style(self, style: Style) -> Self {
        self.state.0.lock().unwrap().style = style;

        self
    }

    /// Sets a function to call when rendering the [`SysTray`].
    pub fn view_callback<F>(self, callback: F) -> Self
    where
        F: Fn(Vec<widget::WidgetDef<Msg>>, &Style) -> Option<widget::WidgetDef<Msg>>
            + Send
            + Sync
            + 'static,
    {
        self.state.0.lock().unwrap().view_callback = Some(Box::new(callback));
        self.emit(widget_signal::RedrawNeeded);

        self
    }

    /// Sets a function to call when rendering the [`SysTray`]'s icons.
    pub fn icon_view_callback<F>(self, callback: F) -> Self
    where
        F: Fn(&status_notifier::Item, IconStyle) -> Option<widget::WidgetDef<Msg>>
            + Send
            + Sync
            + 'static,
    {
        self.state.0.lock().unwrap().icon_view_callback = Some(Box::new(callback));
        self.emit(widget_signal::RedrawNeeded);

        self
    }

    /// Create a [`WeakSysTray`] pointing to the same state.
    pub fn downgrade(&self) -> WeakSysTray<Msg> {
        WeakSysTray(self.state.downgrade())
    }

    /// Default view to render SysTray.
    pub fn default_view(
        children: Vec<widget::WidgetDef<Msg>>,
        style: &Style,
    ) -> Option<widget::WidgetDef<Msg>> {
        let mut row = Row::new_with_children(children)
            .height(Length::Fill)
            .item_alignment(widget::Alignment::Center);

        row.spacing = style.spacing;

        let mut container = Container::new(row).style(container::Style {
            border: style.border,
            background_color: style.bg_color,
            ..Default::default()
        });

        container.padding = style.padding;

        Some(container.into())
    }

    /// Default view to render icons.
    pub fn default_icon_view(
        item: &status_notifier::Item,
        style: IconStyle,
    ) -> Option<widget::WidgetDef<Msg>> {
        let path = item
            .icon_theme_path()
            .zip(item.icon_name())
            .map(|(path, name)| format!("{path}/{name}.png"));

        let handle = if let Some(path) = path {
            image::Handle::Path(path.into())
        } else {
            let handle = item.with_pixmap(|pixmap| {
                let pixmap = pixmap.first()?;

                Some(widget::image::Handle::Rgba {
                    width: pixmap.width as u32,
                    height: pixmap.height as u32,
                    bytes: pixmap.bytes.clone(),
                })
            });

            handle.unwrap_or_else(|| {
                use crate::misc::color;
                use crate::misc::icons;

                icons::misc::broken_picture().to_image_handle(Some(color::from_hex("#FFFFFF")))
            })
        };

        let mut container = Container::new(Image::new(handle)).style(container::Style {
            border: style.border,
            background_color: style.bg_color,
            ..Default::default()
        });

        container.padding = style.padding;

        Some(container.into())
    }

    /// Default configuration used to create menus.
    pub fn default_menu_config() -> MenuConfig {
        MenuConfig {
            direction: Some(crate::menu::Direction::DownLeft),
            popup_config: Some(
                PopupConfig::new()
                    .anchor(popup::Anchor::BottomRight)
                    .offset(popup::Offset { x: 0., y: 8. }),
            ),
            child_popup_config: Some(PopupConfig::new().offset(popup::Offset { x: -2., y: 0. })),
            ..Default::default()
        }
    }
}

impl<Msg> WeakSysTray<Msg> {
    /// Tries to upgrade to a [`SysTray`], returning None if the value has already been destroyed.
    pub fn upgrade(&self) -> Option<SysTray<Msg>> {
        self.0.upgrade().map(|state| SysTray { state })
    }
}

impl<Msg> Inner<Msg>
where
    Msg: Clone + From<WidgetMessage> + Into<Option<WidgetMessage>> + Send + Sync + 'static,
{
    fn close_menu(&mut self) {
        let Some(menu) = self.menu.take() else {
            return;
        };

        if let Some(handle) = self.menu_signal.take() {
            menu.with_emitter().disconnect(handle);
        }

        menu.close();
    }

    fn view_icon(&self, item: &status_notifier::Item) -> Option<widget::WidgetDef<Msg>> {
        let key = item.name();
        let id = item.id();

        let style = self.style.get_icon_style(
            self.active.as_ref() == Some(&id),
            self.hovered.as_ref() == Some(&id),
        );
        let view = if let Some(callback) = &self.icon_view_callback {
            callback(item, style)?
        } else {
            SysTray::default_icon_view(item, style)?
        };

        let mouse_area = MouseArea::new(view)
            .on_release(self.message_builder.activate(key.clone()).into())
            .on_enter(self.message_builder.enter(key.clone()).into())
            .on_exit(self.message_builder.exit(key.clone()).into());

        Some(Container::new(mouse_area).id(id).into())
    }

    fn activate_item(&mut self, name: String, parent: Option<snowcap_api::popup::Parent>) {
        let config = self.menu_config.clone();
        let Some((root, id)) = self.host.with_items(|items| {
            items
                .iter()
                .find(|i| i.name() == name && i.is_menu())
                .and_then(|item| {
                    item.with_layout({
                        let id = item.id();
                        let weak = item.downgrade();
                        move |layout| {
                            let node = layout?;
                            Some((menu::Entry::new(weak, node, config.clone()), id))
                        }
                    })
                })
        }) else {
            return;
        };

        let Some(parent) = parent else {
            tracing::error!("Could not open menu: No parent handle");
            return;
        };

        if let Some(menu) = root.open_menu() {
            self.close_menu();

            let menu = menu.show(
                parent,
                None,
                Some(PopupConfig::new().position(Position::AtWidget(id.clone()))),
            );
            let handle = menu.with_emitter().connect({
                let deactivate = self.message_builder.deactivate(id.clone());
                let emitter = self.with_emitter();

                move |crate::menu::signal::Closed| {
                    emitter.emit(crate::widget::signal::Message::<Msg>(
                        deactivate.clone().into(),
                    ));
                    signal::HandlerPolicy::Discard
                }
            });

            self.menu = Some(menu);
            self.menu_signal = Some(handle);
            self.active = Some(id);
        }
    }

    fn deactivate_item(&mut self, name: String) {
        if self.active == Some(name) {
            self.menu_signal.take();
            self.menu.take();
        }
    }
}

impl MessageBuilder<Action> {
    fn activate(&self, key: String) -> WidgetMessage {
        self.build(Action::Activate(key))
    }

    fn deactivate(&self, key: String) -> WidgetMessage {
        self.build(Action::Deactivate(key))
    }

    fn enter(&self, key: String) -> WidgetMessage {
        self.build(Action::Enter(key))
    }

    fn exit(&self, key: String) -> WidgetMessage {
        self.build(Action::Exit(key))
    }
}

impl<Msg> WithEmitter for Inner<Msg> {
    fn with_emitter(&self) -> crate::signal::Emitter {
        self.base.with_emitter()
    }
}

impl<Msg> WithEmitter for SysTray<Msg> {
    fn with_emitter(&self) -> crate::signal::Emitter {
        self.state.0.lock().unwrap().with_emitter()
    }
}

impl<Msg> Widget for Inner<Msg>
where
    Msg: Clone + From<WidgetMessage> + Into<Option<WidgetMessage>> + Send + Sync + 'static,
{
    type Message = Msg;

    fn view(&self) -> Option<widget::WidgetDef<Self::Message>> {
        let children = self
            .host
            .with_items(|items| items.iter().flat_map(|item| self.view_icon(item)).collect());

        if let Some(callback) = &self.view_callback {
            callback(children, &self.style)
        } else {
            SysTray::default_view(children, &self.style)
        }
    }

    fn update(&mut self, msg: Self::Message, parent: Option<snowcap_api::popup::Parent>) {
        let Some(msg) = msg.into() else {
            return;
        };

        let action = match msg {
            WidgetMessage::SysTray(Message { id, action }) if id == self.base.id() => action,
            _ => return,
        };

        match action {
            Action::Enter(name) => {
                self.hovered = Some(name);
            }
            Action::Exit(name) => {
                if self.hovered == Some(name) {
                    self.hovered = None;
                }
            }
            Action::Activate(name) => self.activate_item(name, parent),
            Action::Deactivate(name) => self.deactivate_item(name),
        }
    }
}

impl<Msg> WithState for SysTray<Msg> {
    type Type = Inner<Msg>;

    fn with_state(&self) -> State<Self::Type> {
        self.state.clone()
    }
}

impl From<Message> for WidgetMessage {
    fn from(value: Message) -> Self {
        Self::SysTray(value)
    }
}
